### ISSUE TITLE
What to do regarding java development casks

### ISSUE THREAD

**Issue by vitorgalvao** [Homebrew/homebrew-cask#57387]
Refs https://github.com/Homebrew/homebrew-cask/pull/57374#issuecomment-453760711. Ping @Homebrew/cask.

I feel like the multitude of Java/JVM/JDK casks is too much. Even which software to call `java` is a contentious subject that pops up once in a while. Those casks need non-trivial amounts of spaghetti code (i.e. non-standard-cask-code), there are a bunch of them, all have slightly different reasons for existing, and more than once we’ve had to rethink how they’re organised. We’ve tried to make sense of them yet again recently, but the person doing most of the work in that regard has since left.

Our options, as I see them:

1. Continue adding these as they’re submitted, increasing the mess, eventually having to remove a bunch of them because nobody can make heads or tails of it anymore, or just leave them hanging and broken, as has happened in the past.
2. Have someone (preferably a core maintainer, but not mandatory) commit to understanding all of these different softwares, why they exist, decide which should be included or not, and keep them in good shape. This person will be made a [code owner](https://blog.github.com/2017-07-06-introducing-code-owners/) to deal with all of these. To improve the [lottery factor](https://en.wikipedia.org/wiki/Bus_factor), this person should also keep a document that keeps track of all these different casks, why they exist, and why they need the extra code to be made functional (particularly useful in cases where they differ, if any).
3. Keep the `java` cask and move everything else to yet another tap.
4. Keep the `java` cask and delete everything else. If everyone else needs them, let them make their own tap.

My takes on these suggestions: **1** looks like a recipe for failure and burnout. Both **3** and **4** might restore order, *but* I’d prefer if that decision would be made by someone with the knowledge described in **2**. **4** will piss off more people, and **3** won’t really ease our burden, just shift it. I’ll definitely not be the person to tackle **2**. I dislike java and the stress java casks have given HBC repos in the past and that would make impartiality in my handling of it difficult. However, I would likely help that person tightening the documentation.

Naturally, any other solutions are welcome, or even opinions if you think this isn’t an issue. Right now, I feel this is a problem, even if an unconscious one. Casks with spaghetti code tend to be left longer without being fixed, java being no exception.

**Comment by amyspark** [comment]
This issue has already cropped up in https://github.com/Homebrew/homebrew-cask/pull/53976.

I've so far removed myself from this sort of discussion, but I feel this is the time to speak up. 

> 4.  Keep the java cask and delete everything else. If everyone else needs them, let them make their own tap.
> 4 will piss off more people,

This is not a matter of how many people will be pissed by our decision (or not), it's a matter of how much will the assigned person last before burning out. When I installed Homebrew (and Cask), I did it because of manageability, and this should also apply to the maintainers.

My vote goes with 4: maintain only official *and* supported versions of Java. Let's be clear that everyone else should host their own taps.

**Comment by vitorgalvao** [comment]
> it's a matter of how much will the assigned person last before burning out.

To clarify, I’m suggesting a volunteer, so someone who *wants* to do it. For example, any user that complains hard about the removal of the casks should instead offer to manage them. I’m not talking about giving them commit access, but they should be expected to make the reviews and fixes and maintain the document (which, if they ever abandon their position, should serve as a guide for the next person).

**Comment by ran-dall** [comment]
> My vote goes with 4: maintain only official and supported versions of Java. Let's be clear that everyone else should host their own taps.

My preference would lay with @amyspark ; and for groups to make their own taps like the [AdoptOpenJDK HomeBrew Tap](https://github.com/AdoptOpenJDK/homebrew-openjdk). Furthermore, we could create a document in the docs to serve as a directory of all known OpenJDK taps; also I think if you are going to create and maintain a tap (like AdoptOpenJDK), you would be responsible for creating... 
> ...a document that keeps track of all (your) different casks (in our tap), why they exist, and why they need the extra code to be made functional (particularly useful in cases where they differ, if any).

**Comment by mjsteinbaugh** [comment]
This is potentially related – I'm seeing a download error for the current version of the java cask:

```
$ brew cask reinstall java
==> Satisfying dependencies
==> Downloading https://download.java.net/java/GA/jdk11/7/GPL/openjdk-11.0.2_osx-x64_bin.t

curl: (22) The requested URL returned error: 404 Not Found
Error: Download failed on Cask 'java' with message: Download failed: https://download.java.net/java/GA/jdk11/7/GPL/openjdk-11.0.2_osx-x64_bin.tar.gz
```

**Comment by vitorgalvao** [comment]
@mjsteinbaugh That is not related. This is an issue to discuss the future of those casks, not to report bugs. Please open a new issue instead, and follow the template.

**Comment by claui** [comment]
Don’t mind me – just collecting a few random factoids to help clear my mind on the subject.

- As of April 2019, 185 formulas on homebrew-core depend on `:java =>` something.

- 40% of those are pinned to 1.8. The others have no upper version bound.

- The Java 8 → 9 version bump [caused significant migration effort](https://github.com/Homebrew/homebrew-core/issues/19696) when additional formulas had to be pinned which had no upper version bound before.

- Losing @ilovezfs as a maintainer didn’t exactly help; see the [9 → 10](https://github.com/Homebrew/homebrew-core/issues/28760) transition.

- Other Java major version bumps, such as [11 → 12](https://github.com/Homebrew/homebrew-cask/pull/60598), caused much less work.

- There should be at least one public, free, long-term supported Java on homebrew-cask at any given moment to satisfy the > 70 core formulas that depend on a pinned Java.

- Java 1.8 is well past its shelf life; [we can’t count on Oracle](https://github.com/Homebrew/homebrew-cask/pull/47609#issuecomment-392289988) to provide a replacement.

- Organizing all Java casks into a separate tap has been considered but it was [discovered that it really wouldn’t help much](https://github.com/Homebrew/homebrew-cask/issues/52875).

- Thanks to @reitermarkus’s work, individual Java casks impose a [lower maintenance tax](https://github.com/Homebrew/homebrew-cask/pull/58510) than they used to. Looks like they scale much better now.

- Being all too exclusionist [can be controversial](https://github.com/Homebrew/homebrew-cask-versions/pull/6453). I don’t fear conflict but ideally, I’d love for the Homebrew org to be on good terms with other key orgs, such as AdoptOpenJDK.

- More and more apps tend to [vendor their own Java](https://github.com/Homebrew/homebrew-core/pull/24508). That won’t help with homebrew-core though, unless we can build that JDK from some OSI-compatible source.

- My design principle for naming Java casks: [Don’t memorize stuff, and don’t design anything that requires others to memorize stuff.](https://github.com/Homebrew/homebrew-cask/issues/41990#issuecomment-363064880)

- One early attempt of mine to document and categorize Java casks: https://github.com/Homebrew/homebrew-cask/pull/54074#issuecomment-432975395 (don’t try this at home.)

- I’d love to become the person to tend to our Java cask garden. (That would be option 2 on the @vitorgalvao scale). But I feel that’s only going to work if I document and automate most of that work in the long run.

**Comment by vitorgalvao** [comment]
> I’d love to become the person to tend to our Java cask garden. (That would be option 2 on the @vitorgalvao scale). But I feel that’s only going to work if I document and automate most of that work in the long run.

Beautiful. Option 2 already predicted such a documentation effort, so that you also see it as pivotal is a good sign it’s an idea with legs.

From my side, I give you carte blanche to proceed however you find best. I reinforce my commitment from the top post of helping you tighten the spots you need.

**Comment by claui** [comment]
Thanks for your support @vitorgalvao. I’ve been procrastinating on writing that documentation so maybe public commitment helps a bit!

**Comment by vikeri** [comment]
Possibly slightly off topic, but [Graal](https://www.graalvm.org/) can compile many JVM applications to a binary that doesn't need the JVM. And if I'm not misunderstanding that could hence remove the need for a Java dependency for those casks and possibly alleviate some of the pain?

**Comment by vitorgalvao** [comment]
@vikeri Wouldn’t solve anything. The issue isn’t casks depending on java, but the amount of casks that *provide* java.

**Comment by vikeri** [comment]
@vitorgalvao I see, totally misunderstood this:
> As of April 2019, 185 formulas on homebrew-core depend on :java => something.

**Comment by claui** [comment]
@vikeri That number may help assess which Java-providing Casks we need to keep.

I should have stated that.

**Comment by vikeri** [comment]
@claui No problem, that's a lot of casks providing java 😬

**Comment by claui** [comment]
@vikeri Just to be on the safe side with ambiguity: those 185 formulas do not provide Java. They depend on Java-providing casks.

**Comment by claui** [comment]
In other news: https://github.com/Homebrew/homebrew-cask-versions/issues/7253

tl;dr Oracle JDK 1.8 seems to have moved behind a registration wall.

**Comment by claui** [comment]
Heads up @vitorgalvao and others: a few Homebrew maintainers have suggested changes. A tad more disruptive than what I had in mind (for now) but I’m fine with it as it makes things simpler.

https://github.com/Homebrew/brew/pull/6035#issuecomment-484927663
https://github.com/Homebrew/brew/pull/6040

Feel free to comment and review (either here or directly in https://github.com/Homebrew/brew/pull/6040). Any input is welcome!

**Comment by vitorgalvao** [comment]
I was already [replying there](https://github.com/Homebrew/brew/pull/6040#issuecomment-484947906), so I kept it.

I’ll add that I’m fine with whichever decision regarding those points, but we do need to decide.

**Comment by MikeMcQuaid** [comment]
My view (just to copy-paste what I was saying in Slack) is that we consider making `brew cask install java` (and even `java8`) install whatever the blessed version of Java we want to support is (which looks like it might be AdoptOpenJDK).

**Comment by tellison** [comment]
We'd be very happy to work with you on ensuring the smooth transition to AdoptOpenJDK.

I've not read every word of the discussion here, but a quick look shows things like @vitorgalvao 's [comment](https://github.com/Homebrew/brew/pull/6040#issuecomment-484947906):
> AdoptOpenJDK has an external tap, meaning we’ll have some duplication efforts. How should 
> we deal with those? If they wish to remain independent but we still want those casks in the main
> repos, I’m willing to build a script/bot that will pull their versions daily.

Again, happy to discuss how to make it easiest for the both of us.  You can see the current state of play in our [HomeBrew Tap repo](https://github.com/AdoptOpenJDK/homebrew-openjdk).

**Comment by karianna** [comment]
> We'd be very happy to work with you on ensuring the smooth transition to AdoptOpenJDK.
> 
> I've not read every word of the discussion here, but a quick look shows things like @vitorgalvao 's [comment](https://github.com/Homebrew/brew/pull/6040#issuecomment-484947906):
> 
> > AdoptOpenJDK has an external tap, meaning we’ll have some duplication efforts. How should
> > we deal with those? If they wish to remain independent but we still want those casks in the main
> > repos, I’m willing to build a script/bot that will pull their versions daily.
> 
> Again, happy to discuss how to make it easiest for the both of us. You can see the current state of play in our [HomeBrew Tap repo](https://github.com/AdoptOpenJDK/homebrew-openjdk).

We originally had an external tap at the request of the Homebrew community.  IIRC they were concerned about the potential clash with the existing Oracle ones and also a large number of casks.

CCing in @gdams as he was closer to it.

**Comment by vitorgalvao** [comment]
Question for the AdoptOpenJDK folks. In the `java` casks we have this simple installation process where we need only [move a single file](https://github.com/Homebrew/homebrew-cask/blob/e22d87d0e34d7463ce55125b1146ffaf796d24b9/Casks/java.rb#L9). [Your method](https://github.com/AdoptOpenJDK/homebrew-openjdk/blob/b08f38ed3c596dd6cf0506dfc33eac1a7cc05820/Casks/adoptopenjdk12.rb) (by the way, there’s no need for an array in `uninstall pkgutil` since it’s a single item) seems to require a `pkg`.

`pkg`s can do anything and write anywhere. An `artifact` is clear and auditable. All else being equal, `artifact` is a net win. Could your installation follow the same method as the `java` casks?

**Comment by yurikoles** [comment]
Oracle now forbids commercial use of their JDK builds without a subscription.
https://www.oracle.com/technetwork/java/java-se-support-roadmap.html

**Comment by karianna** [comment]
> Question for the AdoptOpenJDK folks. In the `java` casks we have this simple installation process where we need only [move a single file](https://github.com/Homebrew/homebrew-cask/blob/e22d87d0e34d7463ce55125b1146ffaf796d24b9/Casks/java.rb#L9). [Your method](https://github.com/AdoptOpenJDK/homebrew-openjdk/blob/b08f38ed3c596dd6cf0506dfc33eac1a7cc05820/Casks/adoptopenjdk12.rb) (by the way, there’s no need for an array in `uninstall pkgutil` since it’s a single item) seems to require a `pkg`.
> 
> `pkg`s can do anything and write anywhere. An `artifact` is clear and auditable. All else being equal, `artifact` is a net win. Could your installation follow the same method as the `java` casks?

Probalby! - I'll defer to @gdams here.  IIRC we use a PKG for two reasons.  One is because as well as brew we also allow users to install purely from a pkg (from our site) and I think we were trying to reduce logic).  The other is that potentially it was easier to install in the correct location and update the Mac OS X internal java linking as well (via the plist).

**Comment by vitorgalvao** [comment]
> we also allow users to install purely from a pkg

For non-Homebrew users it makes sense to have the `pkg`, agreed. But if you could also build the `artifact` version for Homebrew Cask at the same time, that’d be appreciated.

**Comment by claui** [comment]
@gdams If I understand AdoptOpenJDK’s build system correctly, it already releases tarballs along with the pkgs, right?

**Comment by gdams** [comment]
@claui yeah we currently ship both tarballs and pkgs

**Comment by gdams** [comment]
> For non-Homebrew users it makes sense to have the pkg, agreed. But if you could also build the artifact version for Homebrew Cask at the same time, that’d be appreciated.

I'm a little unclear why the extract method is preferred here? Many other casks already use pkg's out there and there are some obvious advantages:

1. You get the added benefit of checking if an application is codesigned.
2. We have far less logic and by shipping pkgs which are generated from a single script, we ensure consistency.
3. We do have some casks with tarballs for older Java versions which we no longer support e.g [here](https://github.com/AdoptOpenJDK/homebrew-openjdk/blob/master/Casks/adoptopenjdk10.rb), but as you can see there is a lot of additional logic required compared to the Oracle casks because we have to change all the plist settings.

In principle, I'm open to either but I do prefer the simplicity and consistency of our PKG casks. My opinion is that we should focus on the more pressing issue first which is that currently there are duplicate casks now if users have tapped both `homebrew/cask-versions` and `AdoptOpenJDK/openjdk`. The way I see it we have 3 options:

1. Migrate all of our casks into `homebrew/cask-versions` and keep the latest `adoptopenjdk` cask in `homebrew-cask`.
2. Remove `adoptopenjdk8` cask from `AdoptOpenJDK/openjdk` to rectify the duplication.
3. Create a new tap at the homebrew project called something like `homebrew/java` and migrate all of our casks to there.

My preference is number 3 as it consolidates all of the code. I'm not sure if homebrew has the facility but it may even make sense to auto "tap" that repo so that users can install Java straight out of the box.

**Comment by vitorgalvao** [comment]
> Many other casks already use pkg's out there

Out of necessity, not because we like it. `pkg`s can do anything and thus are harder to deal with on uninstallations. The only thing worse is `installer`. Every other artifact is a straightforward move (or link, in the case of `binary`) which is simpler to deal with, especially when users have broken installs. Just a few days ago this exact thing happened with these very casks, where a user had multiple versions and wanted just one. “Delete this file”, done.

> but as you can see there is a lot of additional logic

That is a good point.

> older Java versions which we no longer support

Undone by this. If it’s not supported, it won’t get changed and the work is already done. And even then, the older versions could use `pkg` and the new ones `artifact`. There’s no reason to have all casks use the worse method because of consistency.

> I'm open to either but I do prefer the simplicity and consistency of our PKG casks.

The simplicity of `artifact` is exactly *the* reason I’m advocating for it. Our disagreement comes from this single point. Unless you’re referring to the simplicity of *building* them, since I’m referring to the simplicity of dealing with them.

But either way, you are correct there’s a more pressing matter now:

> My opinion is that we should focus on the more pressing issue first which is that currently there are duplicate casks now if users have tapped both `homebrew/cask-versions` and `AdoptOpenJDK/openjdk`.

> My preference is number 3 as it consolidates all of the code.

That’s unlikely to happen (though I wouldn’t oppose it). Homebrew as a whole has been moving away from having more than one official tap, so we’re likely to remove taps, not add them. Option 1 might be best, but I’m not sure we want *all* casks, or just the `jdk`s non-`openj9`.

**Comment by claui** [comment]
> The simplicity of `artifact` is exactly _the_ reason I’m advocating for it. Our disagreement comes from this single point. Unless you’re referring to the simplicity of _building_ them, since I’m referring to the simplicity of dealing with them.

@vitorgalvao There’s also a case for keeping `pkg`s across the board: fewer variants means fewer things that can break for our users, all else being equal. No strong opinion though.

> That’s unlikely to happen (though I wouldn’t oppose it). Homebrew as a whole has been moving away from having more than one official tap, so we’re likely to remove taps, not add them.

I strongly agree. Getting rid of the tap zoo was not only a huge effort from a maintainers’ perspective but also a significant win for users.

> Option 1 might be best,

Strong 👍 from me for this as well.

> but I’m not sure we want _all_ casks, or just the `jdk`s non-`openj9`.

I’m pretty sure we want them all.

Homebrew users may appreciate having a single source, not a mixture of taps.
The folks at AdoptOpenJDK may appreciate no longer having to maintain a custom tap.
And from a Homebrew Cask maintainer’s perspective, it may not even matter much because given all that has been said, I’d like for us to automate (Homebrew’s part of) the Java cask release pipeline anyway.

**Comment by gdams** [comment]
> but I’m not sure we want all casks, or just the jdks non-openj9.

I think that the AdoptOpenJDK project would massively prefer to have them all in once place.

**Comment by gdams** [comment]
> I’d like for us to automate (Homebrew’s part of) the Java cask release pipeline anyway.

we have a good API that could be used to help with this.

**Comment by claui** [comment]
> > but I’m not sure we want all casks, or just the jdks non-openj9.
> 
> I think that the AdoptOpenJDK project would massively prefer to have them all in once place.

@gdams Got it.

By the way, would it make sense to drop both the Java 9 and 10 cask immediately?
Both are obviously [unsupported](https://adoptopenjdk.net/support.html) and the associated builds [haven’t been updated](https://github.com/AdoptOpenJDK/openjdk9-binaries/releases) for [18 months](https://github.com/AdoptOpenJDK/openjdk10-binaries/releases).

Users who are not aware of the LTS three-step dance may inadvertently install insecure builds, right?

**Comment by claui** [comment]
> we have a good API that could be used to help with this.

@gdams Excellent. I’d love to tackle that right after the initial migration.

@vitorgalvao It would be great to keep the PR count as low as possible. Would you be OK with making AdoptOpenJDK updates (not additions) exempt from our one-cask-per-PR rule?

**Comment by gdams** [comment]
> Would you be OK with making AdoptOpenJDK updates (not additions) exempt from our one-cask-per-PR rule?

This would help reduce a lot of clutter.

**Comment by claui** [comment]
My suggestion would be to keep the following casks:

## Java casks, April 2019

| Major version   | Supported? | JDK HotSpot | JDK OpenJ9 | JRE HotSpot | JRE OpenJ9 |
|:--------------- |:----------:|:-----------:|:----------:|:-----------:|:----------:|
|  **8 (LTS)**    | At least until 2023‑09 | In `cask‑version` | In `cask‑version` | In `cask‑version` | In `cask‑version` |
| **11 (LTS)**    | At least until 2022‑09 | In `cask‑version` | In `cask‑version` | In `cask‑version` | In `cask‑version` |
| **12 (latest)** | At least until 2022‑09 | **In `cask`<br/>(unnumbered)** | **In `cask‑version`<br/>(unnumbered)** | **In `cask‑version`<br/>(unnumbered)** | **In `cask‑version`<br/>(unnumbered)** |

I recommend that unsupported casks be pruned, maybe with a grace period of six months. That way, users still have a fallback in case the latest release breaks things for them.

Here’s how the Java cask zoo would look five years from now, in April 2024:

## Java casks, April 2024, with pruning

| Major version   | Supported? | JDK HotSpot | JDK OpenJ9 | JRE HotSpot | JRE OpenJ9 |
|:--------------- |:----------:|:-----------:|:----------:|:-----------:|:----------:|
|  **8 (LTS)**    | Maybe      | In `cask‑version` (maybe) | In `cask‑version` (maybe) | In `cask‑version` (maybe) | In `cask‑version` (maybe) |
| **11 (LTS)**    | Unlikely   | —          | —          | —          | —          |
| **14 (LTS)**    | Maybe      | In `cask‑version` (maybe) | In `cask‑version` (maybe) | In `cask‑version` (maybe) | In `cask‑version` (maybe) |
| **17 (LTS)**    | Until 2025‑09 (I suppose) | In `cask‑version` | In `cask‑version` | In `cask‑version` | In `cask‑version` |
| **20 (LTS)**    | Until 2027‑03 (I suppose) | In `cask‑version` | In `cask‑version` | In `cask‑version` | In `cask‑version` |
| **21**          | —          | In `cask-version` but freshly EOL | —          | —          | —          |
| **22 (latest)** | Until 2024‑09 (I suppose) | **In `cask`<br/>(unnumbered)** | **In `cask‑version`<br/>(unnumbered)** | **In `cask‑version`<br/>(unnumbered)** | **In `cask‑version`<br/>(unnumbered)** |

I think that scheme would be sustainable.

**Comment by dwijnand** [comment]
LTS releases are every 3 years, not every 3 releases, and given a release is every 6-months that means it's an LTS release every 6 releases.

So... Java 14 and Java 20 aren't LTS releases.  The next LTS after Java 11 is Java 17, in 2023-09.

**Comment by vitorgalvao** [comment]
As a note, I think version `8` and `12` that we now have are different. `pkg` on the former, `artifact` on the latter.

> Would you be OK with making AdoptOpenJDK updates (not additions) exempt from our one-cask-per-PR rule?

Travis doesn’t run with more than one cask, so ironically more PRs means less work for us, because they can be automerged.

But I’m just pointing out a practical consideration; I have no objection with the single-PR method for these casks if it saves trouble somewhere else.

**Comment by claui** [comment]
> LTS releases are every 3 years, not every 3 releases

Thanks! Fixed.

**Comment by claui** [comment]
> Travis doesn’t run with more than one cask, so ironically more PRs means less work for us, because they can be automerged.

Good point @vitorgalvao. I didn’t think of that.

> But I’m just pointing out a practical consideration; I have no objection with the single-PR method for these casks if it saves trouble somewhere else.

I’m fine with the single-PR method (for now) unless @gdams or @karianna object.

**Comment by karianna** [comment]
No objection from me
Cheers,
Martijn

**Comment by vitorgalvao** [comment]
A point I forgot to make earlier. We have several Java-providing casks, and they all use `artifact` instead of `pkg`. Maintaining that consistency is another point for ease of maintainability on our side.

**Comment by mfriedenhagen** [comment]
What I would expect as a user:
* adoptopenjdk in casks always install the latest major version.
* LTS versions are all available via cask-versions as adoptopenjdkMAJOR, i.e. adoptopenjdk8 and adoptopenjdk11.
* Non-LTS versions except of the latest are available via the adoptopenjdk tap, i.e. 9, 10 and 12 are available as adoptopenjdk9, adoptopenjdk10 and adoptopenjdk12.

Then you have a clear separation (well, at least I would be able to grasp this :-)).

**Comment by ryenus** [comment]
It seems we already have a conclusion, which is to [endorse **adoptopenjdk** as the way to go, as @claui [stated in another thread back in Apr 2019][1]

> **tl;dr We’re planning to endorse AdoptOpenJDK only.**

That means the java cask, as well as the recent java versions from cask-versions should be all updated to point the builds from adoptopenjdk, right?

It's a pity that the java versions from cask-versions is still a bit messy as of today.

[1]: https://github.com/Homebrew/brew/pull/6035#issuecomment-484927663

**Comment by reitermarkus** [comment]
> That means the java cask, as well as the recent java versions from cask-versions should be all updated to point to the builds from adoptopenjdk, right?

No, this only means we will suggest installing `adoptopenjdk` instead of `java`. This was done in https://github.com/Homebrew/brew/pull/6472.

**Comment by ryenus** [comment]
@reitermarkus, thank you for the clarification. I fully agree with the preference to **adoptopenjdk**.

Given that we're keeping the name `java` for the latest release of what is now called `openjdk`, I think it should be fine to use `java{8,9.10,11,12}` (could be as aliases) to refer to previous versions of the builds from our preferred vendor.

I understand we need to deliver the message about the change of java vendor, but usually it's better to educate the users with clear communication, rather than breaking things.

Another problem is, currently in `cask-versions`, there's `adoptopenjdk8`, but the other (newer) versions are missing. This is problematic because to install newer adoptopenjdk releases, one needs to tap `adoptopenjdk/openjdk`, which also provides adoptopenjdk8, causing a conflict with cask-versions on adoptopenjdk8.

This indeed has been discussed before, and it's been awhile, maybe it's a good time for a re-examination?

**Comment by reitermarkus** [comment]
> could be as aliases

Casks don't support aliases (yet). Even if they did, I think aliasing `java` to `adoptopenjdk` is more confusing than helpful.

> causing a conflict with cask-versions on adoptopenjdk8

I think there's an open issue for this bug. If there isn't one, someone should open one, or better yet, submit a PR fixing it.

**Comment by vitorgalvao** [comment]
Our current Java-related casks are considerably simpler than they were at the time this issue was opened. In other words, the situation has changed. This issue was relevant when the casks were a huge maintenance burden and that is no longer the case.

I propose we resume accepting submissions for those casks, **provided they work well without workarounds**. That means passing CI without [`*flight` blocks](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/flight.md) or [`caveats`](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/caveats.md) with instructions for manual steps. [What we have](https://github.com/Homebrew/homebrew-cask/blob/e697ce1dd5f3131d251c2a2ab4579cd628158315/Casks/adoptopenjdk.rb) is acceptable; [what we had](https://github.com/Homebrew/homebrew-cask/blob/a7d937d23e5c644a4806614241eb7fb4db09f53b/Casks/adoptopenjdk.rb) isn’t. Any cask of this kind which is deemed to be too complex can and should be rejected.

---

#### Important considerations

* At this point, *this is a recommendation*. Other maintainers who disagree can block it.
* If this turns out to be a bad decision, we will revert it and may delete the new casks.
* We should take it slow. Please don’t start submitting these en masse or multiple versions to [homebrew/cask-versions](https://github.com/Homebrew/homebrew-cask-versions). Let’s start with the main repo first and see how it goes.
* Don’t use currently-included casks as examples of “that one does it, so why can’t this one”. Removing an included casks causes more disruption than refusing one that isn’t yet in. The former get more leeway.
* If a maintainer refuses one of these, please don’t be difficult. If we go through with allowing these casks again, it’s in everyone’s best interest we don’t regret the decision.

**Comment by vitorgalvao** [comment]
No objections were raised, so we’ll go with [the plan](https://github.com/Homebrew/homebrew-cask/issues/57387#issuecomment-766936954).
